File: Se7enPro/Models/UserSettings.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Se7enPro.Models;

public sealed class UserSettings
{
    [JsonPropertyName("theme")]
    public string Theme { get; set; } = "dark";

    [JsonPropertyName("language")]
    public string Language { get; set; } = "en";

    [JsonPropertyName("egressRegion")]
    public string EgressRegion { get; set; } = "";

    [JsonPropertyName("disableTimeouts")]
    public bool DisableTimeouts { get; set; }

    public const int MinPort = 1;
    public const int MaxPort = 65535;

    public static int SanitizePort(int port, int fallback = 0)
        => port is >= MinPort and <= MaxPort ? port : fallback;

    private int _localSocksProxyPort;
    [JsonPropertyName("localSocksProxyPort")]
    public int LocalSocksProxyPort
    {
        get => _localSocksProxyPort;
        set => _localSocksProxyPort = SanitizePort(value);
    }

    private int _localHttpProxyPort;
    [JsonPropertyName("localHttpProxyPort")]
    public int LocalHttpProxyPort
    {
        get => _localHttpProxyPort;
        set => _localHttpProxyPort = SanitizePort(value);
    }

    [JsonPropertyName("useCustomProxyPorts")]
    public bool UseCustomProxyPorts { get; set; }

    [JsonPropertyName("allowLanConnections")]
    public bool AllowLanConnections { get; set; }

    [JsonPropertyName("setSystemProxy")]
    public bool SetSystemProxy { get; set; } = true;

    [JsonPropertyName("autoConnect")]
    public bool AutoConnect { get; set; }

    [JsonPropertyName("startWithWindows")]
    public bool StartWithWindows { get; set; }

    [JsonPropertyName("ipScannerEnabled")]
    public bool IpScannerEnabled { get; set; }

    [JsonPropertyName("minimizeToTray")]
    public bool MinimizeToTray { get; set; } = true;

    [JsonPropertyName("onCloseAction")]
    public string OnCloseAction { get; set; } = "ask";

    [JsonPropertyName("upstreamProxy")]
    public string UpstreamProxy { get; set; } = "";

    [JsonPropertyName("upstreamProxyEnabled")]
    public bool UpstreamProxyEnabled { get; set; } = false;

    [JsonPropertyName("upstreamProxyScheme")]
    public string UpstreamProxyScheme { get; set; } = "http";

    [JsonPropertyName("upstreamProxyUsername")]
    public string UpstreamProxyUsername { get; set; } = "";

    [JsonPropertyName("upstreamProxyPassword")]
    [System.Text.Json.Serialization.JsonConverter(typeof(Services.EncryptedStringConverter))]
    public string UpstreamProxyPassword { get; set; } = "";

    [JsonPropertyName("systemWideTunneling")]
    public bool SystemWideTunneling { get; set; }

    [JsonPropertyName("killSwitchEnabled")]
    public bool KillSwitchEnabled { get; set; }

    [JsonPropertyName("protocolMode")]
    public string ProtocolMode { get; set; } = "auto";

    [JsonPropertyName("beastMode")]
    public bool BeastMode { get; set; }

    [JsonPropertyName("cdnFrontingCustomIpList")]
    public string CdnFrontingCustomIpList { get; set; } = "";

    [JsonPropertyName("cdnFrontingCustomSni")]
    public string CdnFrontingCustomSni { get; set; } = "";

    [JsonPropertyName("cdnFrontingSkipCertVerify")]
    public bool CdnFrontingSkipCertVerify { get; set; } = false;

    [JsonPropertyName("autoFindIpAndSni")]
    public bool AutoFindIpAndSni { get; set; }

    [JsonPropertyName("saveFoundIpsAndSni")]
    public bool SaveFoundIpsAndSni { get; set; } = false;

    [JsonPropertyName("frontedMeekCDNScanBuiltInSets")]
    public List<string> FrontedMeekCDNScanBuiltInSets { get; set; } = new() { "psiphon-akamai", "fastly" };

    [JsonPropertyName("establishTunnelTimeoutSeconds")]
    public int? EstablishTunnelTimeoutSeconds { get; set; } = 300;

    [JsonPropertyName("conduitMode")]
    public string ConduitMode { get; set; } = "auto";

    [JsonPropertyName("conduitCompartmentId")]
    public string ConduitCompartmentId { get; set; } = "";

    [JsonPropertyName("conduitRejectCensoredCountries")]
    public bool ConduitRejectCensoredCountries { get; set; } = true;

    [JsonPropertyName("lanAuthEnabled")]
    public bool LanAuthEnabled { get; set; } = false;

    [JsonPropertyName("lanProxyUsername")]
    public string LanProxyUsername { get; set; } = "";

    [JsonPropertyName("lanProxyPassword")]
    [System.Text.Json.Serialization.JsonConverter(typeof(Services.EncryptedStringConverter))]
    public string LanProxyPassword { get; set; } = "";

    [JsonPropertyName("connectionMethod")]
    public string ConnectionMethod { get; set; } = "wireguard";

    [JsonPropertyName("aetherProtocol")]
    public string AetherProtocol { get; set; } = "masque";

    [JsonPropertyName("aetherEndpointMasque")]
    public string AetherEndpointMasque { get; set; } = "";

    [JsonPropertyName("aetherEndpointWireguard")]
    public string AetherEndpointWireguard { get; set; } = "";

    [JsonPropertyName("aetherEndpointWarp")]
    public string AetherEndpointWarp { get; set; } = "";

    [JsonPropertyName("aetherEndpointMasqueOnMasque")]
    public string AetherEndpointMasqueOnMasque { get; set; } = "";

    [JsonPropertyName("aetherScanMode")]
    public string AetherScanMode { get; set; } = "balanced";

    [JsonPropertyName("aetherNoize")]
    public string AetherNoize { get; set; } = "balanced";

    [JsonPropertyName("aetherIpVersion")]
    public string AetherIpVersion { get; set; } = "4";

    [JsonPropertyName("aetherPerfProfile")]
    public string AetherPerfProfile { get; set; } = "auto";

    [JsonPropertyName("aetherManualPeer")]
    public string AetherManualPeer { get; set; } = "";

    [JsonPropertyName("chainedOuterTransport")]
    public string ChainedOuterTransport { get; set; } = "auto";

    [JsonPropertyName("chainedPsiphonOuterTransport")]
    public string ChainedPsiphonOuterTransport { get; set; } = "auto";

    [JsonPropertyName("chainedTorOuterTransport")]
    public string ChainedTorOuterTransport { get; set; } = "auto";

    [JsonPropertyName("chainedSubMode")]
    public string ChainedSubMode { get; set; } = "psiphon_warp";

    [JsonPropertyName("shadowsocksMethod")]
    public string ShadowsocksMethod { get; set; } = "2022-blake3-aes-128-gcm";

    [JsonPropertyName("aetherEch")]
    public bool AetherEch { get; set; }

    [JsonPropertyName("aetherMasqueTransport")]
    public string AetherMasqueTransport { get; set; } = "h3";

    [JsonPropertyName("aetherFragment")]
    public bool AetherFragment { get; set; } = false;

    [JsonPropertyName("aetherMimEnabled")]
    public bool AetherMimEnabled { get; set; } = false;

    [JsonPropertyName("aetherMimOuter")]
    public string AetherMimOuter { get; set; } = "";

    [JsonPropertyName("aetherMimInner")]
    public string AetherMimInner { get; set; } = "";

    [JsonPropertyName("aetherMimPeers")]
    public string AetherMimPeers { get; set; } = "";

    [JsonPropertyName("aetherQuicV2")]
    public bool AetherQuicV2 { get; set; } = true;

    [JsonPropertyName("aetherMark")]
    public string AetherMark { get; set; } = "";

    [JsonPropertyName("torExitCountry")]
    public string TorExitCountry { get; set; } = "";

    [JsonPropertyName("torBridges")]
    public string TorBridges { get; set; } = "";

    [JsonPropertyName("splitTunnelEnabled")]
    public bool SplitTunnelEnabled { get; set; }

    [JsonPropertyName("splitTunnelMode")]
    public string SplitTunnelMode { get; set; } = "exclude";

    private List<SplitTunnelEntry> _splitTunnelEntries = new();
    [JsonPropertyName("splitTunnelEntries")]
    public List<SplitTunnelEntry> SplitTunnelEntries
    {
        get => _splitTunnelEntries ??= new();
        set => _splitTunnelEntries = value ?? new();
    }

    [JsonPropertyName("v2rayCore")]
    public string V2RayCore { get; set; } = "xray";

    private int _v2rayInboundPort = 10808;
    [JsonPropertyName("v2rayInboundPort")]
    public int V2RayInboundPort
    {
        get => _v2rayInboundPort;
        set => _v2rayInboundPort = SanitizePort(value, 10808);
    }

    [JsonPropertyName("v2rayEnableMux")]
    public bool V2RayEnableMux { get; set; } = false;

    [JsonPropertyName("v2rayRouteDnsThroughV2Ray")]
    public bool V2RayRouteDnsThroughV2Ray { get; set; } = true;

    [JsonPropertyName("v2rayEnableFragment")]
    public bool V2RayEnableFragment { get; set; } = true;

    [JsonPropertyName("v2rayFragmentPackets")]
    public string V2RayFragmentPackets { get; set; } = "tlshello";

    [JsonPropertyName("v2rayFragmentLength")]
    public string V2RayFragmentLength { get; set; } = "100-200";

    [JsonPropertyName("v2rayFragmentInterval")]
    public string V2RayFragmentInterval { get; set; } = "10-20";

    [JsonPropertyName("v2rayActiveConfigId")]
    public string V2RayActiveConfigId { get; set; } = "";

    private List<V2RayConfigEntry> _v2rayConfigs = new();
    [JsonPropertyName("v2rayConfigs")]
    public List<V2RayConfigEntry> V2RayConfigs
    {
        get => _v2rayConfigs ??= new();
        set => _v2rayConfigs = value ?? new();
    }

    [JsonPropertyName("shardCustomCfIp")]
    public string ShardCustomCfIp { get; set; } = "";

    [JsonPropertyName("shardSmartSplit")]
    public bool ShardSmartSplit { get; set; } = false;

    [JsonPropertyName("shardRotateIp")]
    public bool ShardRotateIp { get; set; } = true;
}

public sealed class V2RayConfigEntry
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("protocol")]
    public string Protocol { get; set; } = "vless";

    [JsonPropertyName("address")]
    public string Address { get; set; } = "";

    [JsonPropertyName("port")]
    public int Port { get; set; } = 443;

    [JsonPropertyName("userId")]
    public string UserId { get; set; } = "";

    [JsonPropertyName("security")]
    public string Security { get; set; } = "reality";

    [JsonPropertyName("network")]
    public string Network { get; set; } = "tcp";

    [JsonPropertyName("path")]
    public string? Path { get; set; }

    [JsonPropertyName("host")]
    public string? Host { get; set; }

    [JsonPropertyName("sni")]
    public string? Sni { get; set; }

    [JsonPropertyName("publicKey")]
    public string? PublicKey { get; set; }

    [JsonPropertyName("shortId")]
    public string? ShortId { get; set; }

    [JsonPropertyName("flow")]
    public string? Flow { get; set; }

    [JsonPropertyName("enableFragment")]
    public bool? EnableFragment { get; set; }

    [JsonPropertyName("isActive")]
    public bool IsActive { get; set; }

    [JsonPropertyName("latency")]
    public int Latency { get; set; } = -1;
}

public sealed class SplitTunnelEntry
{
    [JsonPropertyName("kind")]
    public string Kind { get; set; } = "domain";

    [JsonPropertyName("value")]
    public string Value { get; set; } = "";
}

File: Se7enFlutter/lib/core/models/user_settings.dart
import 'dart:convert';

class SplitTunnelEntry {
  SplitTunnelEntry({this.kind = 'domain', this.value = ''});

  String kind;
  String value;

  factory SplitTunnelEntry.fromJson(Map<String, dynamic> j) => SplitTunnelEntry(
        kind: j['kind'] as String? ?? 'domain',
        value: j['value'] as String? ?? '',
      );

  Map<String, dynamic> toJson() => {'kind': kind, 'value': value};

  SplitTunnelEntry copy() => SplitTunnelEntry(kind: kind, value: value);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SplitTunnelEntry &&
          runtimeType == other.runtimeType &&
          kind.toLowerCase() == other.kind.toLowerCase() &&
          value.trim().toLowerCase() == other.value.trim().toLowerCase();

  @override
  int get hashCode => Object.hash(kind.toLowerCase(), value.trim().toLowerCase());
}

class UserSettings {

  String theme;
  String language;

  String egressRegion;
  bool disableTimeouts;
  int localSocksProxyPort;
  int localHttpProxyPort;
  bool useCustomProxyPorts;
  bool allowLanConnections;
  bool setSystemProxy;
  bool autoConnect;
  bool startWithWindows;
  bool ipScannerEnabled;
  bool minimizeToTray;
  String onCloseAction;

  String upstreamProxy;
  bool upstreamProxyEnabled;
  String upstreamProxyScheme;
  String upstreamProxyUsername;
  String upstreamProxyPassword;

  bool systemWideTunneling;
  bool killSwitchEnabled;

  String protocolMode;
  bool beastMode;
  String cdnFrontingCustomIpList;
  String cdnFrontingCustomSni;
  bool cdnFrontingSkipCertVerify;
  bool autoFindIpAndSni;
  bool saveFoundIpsAndSni;
  List<String> frontedMeekCDNScanBuiltInSets;
  int? establishTunnelTimeoutSeconds;

  String conduitMode;
  String conduitCompartmentId;
  bool conduitRejectCensoredCountries;

  bool lanAuthEnabled;
  String lanProxyUsername;
  String lanProxyPassword;

  String connectionMethod;

  String aetherProtocol;
  String aetherScanMode;
  String aetherNoize;
  String aetherIpVersion;
  String aetherManualPeer;
  String aetherEndpointMasque;
  String aetherEndpointWireguard;
  String aetherEndpointWarp;
  String aetherEndpointMasqueOnMasque;
  String chainedOuterTransport;
  String chainedPsiphonOuterTransport;
  String chainedTorOuterTransport;
  String chainedSubMode;
  String aetherMasqueTransport;
  bool aetherFragment;
  String aetherPerfProfile;
  bool aetherMimEnabled;
  String aetherMimOuter;
  String aetherMimInner;
  String aetherMimPeers;
  bool aetherQuicV2;
  String aetherMark;

  String torExitCountry;
  String torBridges;

  bool splitTunnelEnabled;
  String splitTunnelMode;
  List<SplitTunnelEntry> splitTunnelEntries;

  String v2rayCore;
  int v2rayInboundPort;
  bool v2rayEnableMux;
  bool v2rayRouteDnsThroughV2Ray;
  bool v2rayEnableFragment;
  String v2rayFragmentPackets;
  String v2rayFragmentLength;
  String v2rayFragmentInterval;
  String v2rayActiveConfigId;
  List<V2RayConfigEntry> v2rayConfigs;

  String shardCustomCfIp;
  bool shardSmartSplit;
  bool shardRotateIp;

  UserSettings({
    this.theme = 'dark',
    this.language = 'en',
    this.egressRegion = '',
    this.disableTimeouts = false,
    this.localSocksProxyPort = 0,
    this.localHttpProxyPort = 0,
    this.useCustomProxyPorts = false,
    this.allowLanConnections = false,
    this.setSystemProxy = true,
    this.autoConnect = false,
    this.startWithWindows = false,
    this.ipScannerEnabled = false,
    this.minimizeToTray = true,
    this.onCloseAction = 'ask',
    this.upstreamProxy = '',
    this.upstreamProxyEnabled = false,
    this.upstreamProxyScheme = 'http',
    this.upstreamProxyUsername = '',
    this.upstreamProxyPassword = '',
    this.systemWideTunneling = false,
    this.killSwitchEnabled = false,
    this.protocolMode = 'auto',
    this.beastMode = false,
    this.cdnFrontingCustomIpList = '',
    this.cdnFrontingCustomSni = '',
    this.cdnFrontingSkipCertVerify = false,
    this.autoFindIpAndSni = false,
    this.saveFoundIpsAndSni = false,
    List<String>? frontedMeekCDNScanBuiltInSets,
    this.establishTunnelTimeoutSeconds = 300,
    this.conduitMode = 'auto',
    this.conduitCompartmentId = '',
    this.conduitRejectCensoredCountries = true,
    this.lanAuthEnabled = false,
    this.lanProxyUsername = '',
    this.lanProxyPassword = '',
    this.connectionMethod = 'wireguard',
    this.aetherProtocol = 'masque',
    this.aetherScanMode = 'balanced',
    this.aetherNoize = 'balanced',
    this.aetherIpVersion = '4',
    this.aetherManualPeer = '',
    this.aetherEndpointMasque = '',
    this.aetherEndpointWireguard = '',
    this.aetherEndpointWarp = '',
    this.aetherEndpointMasqueOnMasque = '',
    this.chainedOuterTransport = 'auto',
    this.chainedPsiphonOuterTransport = 'auto',
    this.chainedTorOuterTransport = 'auto',
    this.chainedSubMode = 'psiphon_warp',
    this.aetherMasqueTransport = 'h3',
    this.aetherFragment = false,
    this.aetherPerfProfile = 'auto',
    this.aetherMimEnabled = false,
    this.aetherMimOuter = '',
    this.aetherMimInner = '',
    this.aetherMimPeers = '',
    this.aetherQuicV2 = true,
    this.aetherMark = '',
    this.torExitCountry = '',
    this.torBridges = '',
    this.splitTunnelEnabled = false,
    this.splitTunnelMode = 'exclude',
    List<SplitTunnelEntry>? splitTunnelEntries,
    this.v2rayCore = 'xray',
    this.v2rayInboundPort = 10808,
    this.v2rayEnableMux = false,
    this.v2rayRouteDnsThroughV2Ray = true,
    this.v2rayEnableFragment = true,
    this.v2rayFragmentPackets = 'tlshello',
    this.v2rayFragmentLength = '100-200',
    this.v2rayFragmentInterval = '10-20',
    this.v2rayActiveConfigId = '',
    List<V2RayConfigEntry>? v2rayConfigs,
    this.shardCustomCfIp = '',
    this.shardSmartSplit = false,
    this.shardRotateIp = true,
  })  : frontedMeekCDNScanBuiltInSets =
              frontedMeekCDNScanBuiltInSets == null
                  ? ['psiphon-akamai', 'fastly']
                  : List<String>.from(frontedMeekCDNScanBuiltInSets),
        splitTunnelEntries = splitTunnelEntries ?? [],
        v2rayConfigs = v2rayConfigs ?? [];

  static const int minPort = 1;
  static const int maxPort = 65535;

  static int sanitizePort(dynamic v, {int fallback = 0}) {
    if (v is! int) return fallback;
    return (v >= minPort && v <= maxPort) ? v : fallback;
  }

  static List<String> _parseStringList(dynamic v) {
    if (v is List) return v.map((e) => e.toString()).where((s) => s.isNotEmpty).toList();
    return [];
  }

  static String _normLang(String? v) {
    final t = (v ?? 'en').trim().toLowerCase();
    if (t == 'ru') return 'ru';
    if (t == 'zh') return 'zh';
    if (t == 'fa') return 'fa';
    return 'en';
  }

  factory UserSettings.fromJson(Map<String, dynamic> j) => UserSettings(
        theme: j['theme'] as String? ?? 'dark',
        language: _normLang(j['language'] as String?),
        egressRegion: j['egressRegion'] as String? ?? '',
        disableTimeouts: j['disableTimeouts'] as bool? ?? false,
        localSocksProxyPort: sanitizePort(j['localSocksProxyPort']),
        localHttpProxyPort: sanitizePort(j['localHttpProxyPort']),
        useCustomProxyPorts: j['useCustomProxyPorts'] as bool? ??
            ((j['localSocksProxyPort'] != null && (j['localSocksProxyPort'] as int? ?? 0) > 0) ||
                (j['localHttpProxyPort'] != null && (j['localHttpProxyPort'] as int? ?? 0) > 0)),
        allowLanConnections: j['allowLanConnections'] as bool? ?? false,
        setSystemProxy: j['setSystemProxy'] as bool? ?? true,
        autoConnect: j['autoConnect'] as bool? ?? false,
        startWithWindows: j['startWithWindows'] as bool? ?? false,
        ipScannerEnabled: j['ipScannerEnabled'] as bool? ?? false,
        minimizeToTray: j['minimizeToTray'] as bool? ?? true,
        onCloseAction: j['onCloseAction'] as String? ?? 'ask',
        upstreamProxy: j['upstreamProxy'] as String? ?? '',
        upstreamProxyEnabled: j['upstreamProxyEnabled'] as bool? ?? false,
        upstreamProxyScheme: j['upstreamProxyScheme'] as String? ?? 'http',
        upstreamProxyUsername: j['upstreamProxyUsername'] as String? ?? '',
        upstreamProxyPassword: j['upstreamProxyPassword'] as String? ?? '',
        systemWideTunneling: j['systemWideTunneling'] as bool? ?? false,
        killSwitchEnabled: j['killSwitchEnabled'] as bool? ?? false,
        protocolMode: j['protocolMode'] as String? ?? 'auto',
        beastMode: j['beastMode'] as bool? ?? false,
        cdnFrontingCustomIpList: j['cdnFrontingCustomIpList'] as String? ?? '',
        cdnFrontingCustomSni: j['cdnFrontingCustomSni'] as String? ?? '',
        cdnFrontingSkipCertVerify:
            j['cdnFrontingSkipCertVerify'] as bool? ?? false,
        autoFindIpAndSni: j['autoFindIpAndSni'] as bool? ?? false,
        saveFoundIpsAndSni: j['saveFoundIpsAndSni'] as bool? ?? false,
        frontedMeekCDNScanBuiltInSets: j.containsKey('frontedMeekCDNScanBuiltInSets')
            ? _parseStringList(j['frontedMeekCDNScanBuiltInSets'])
            : ['psiphon-akamai', 'fastly'],
        establishTunnelTimeoutSeconds: j['establishTunnelTimeoutSeconds'] is int
            ? j['establishTunnelTimeoutSeconds'] as int
            : (j['establishTunnelTimeoutSeconds'] is num
                ? (j['establishTunnelTimeoutSeconds'] as num).toInt()
                : 300),
        conduitMode: j['conduitMode'] as String? ?? 'auto',
        conduitCompartmentId: j['conduitCompartmentId'] as String? ?? '',
        conduitRejectCensoredCountries:
            j['conduitRejectCensoredCountries'] as bool? ?? true,
        lanAuthEnabled: j['lanAuthEnabled'] as bool? ?? false,
        lanProxyUsername: j['lanProxyUsername'] as String? ?? '',
        lanProxyPassword: j['lanProxyPassword'] as String? ?? '',
        connectionMethod: j['connectionMethod'] as String? ?? 'wireguard',
        aetherProtocol: j['aetherProtocol'] as String? ?? 'masque',
        aetherScanMode: j['aetherScanMode'] as String? ?? 'balanced',
        aetherNoize: j['aetherNoize'] as String? ?? 'balanced',
        aetherIpVersion: j['aetherIpVersion'] as String? ?? '4',
        aetherManualPeer: j['aetherManualPeer'] as String? ?? '',
        aetherEndpointMasque: j['aetherEndpointMasque'] as String? ?? '',
        aetherEndpointWireguard: j['aetherEndpointWireguard'] as String? ?? '',
        aetherEndpointWarp: j['aetherEndpointWarp'] as String? ?? '',
        aetherEndpointMasqueOnMasque:
            j['aetherEndpointMasqueOnMasque'] as String? ?? '',
        chainedOuterTransport: j['chainedOuterTransport'] as String? ?? 'auto',
        chainedPsiphonOuterTransport: (j['chainedPsiphonOuterTransport'] as String?)?.trim().isNotEmpty == true
            ? (j['chainedPsiphonOuterTransport'] as String).trim()
            : ((j['chainedOuterTransport'] as String?)?.trim().isNotEmpty == true ? (j['chainedOuterTransport'] as String).trim() : 'auto'),
        chainedTorOuterTransport:
            (j['chainedTorOuterTransport'] as String?)?.trim().isNotEmpty == true
                ? (j['chainedTorOuterTransport'] as String).trim()
                : ((j['chainedOuterTransport'] as String?)?.trim().isNotEmpty == true ? (j['chainedOuterTransport'] as String).trim() : 'auto'),

        chainedSubMode: j['chainedSubMode'] as String? ?? 'psiphon_warp',
        aetherMasqueTransport: j['aetherMasqueTransport'] as String? ?? 'h3',
        aetherFragment: j['aetherFragment'] as bool? ?? false,
        aetherPerfProfile: j['aetherPerfProfile'] as String? ?? 'auto',
        aetherMimEnabled: j['aetherMimEnabled'] as bool? ?? false,
        aetherMimOuter: j['aetherMimOuter'] as String? ?? '',
        aetherMimInner: j['aetherMimInner'] as String? ?? '',
        aetherMimPeers: j['aetherMimPeers'] as String? ?? '',
        aetherQuicV2: j['aetherQuicV2'] as bool? ?? true,
        aetherMark: j['aetherMark'] as String? ?? '',
        torExitCountry: j['torExitCountry'] as String? ?? '',
        torBridges: j['torBridges'] as String? ?? '',
        splitTunnelEnabled: j['splitTunnelEnabled'] as bool? ?? false,
        splitTunnelMode: j['splitTunnelMode'] as String? ?? 'exclude',
        splitTunnelEntries: (j['splitTunnelEntries'] as List<dynamic>?)
                ?.map((e) => SplitTunnelEntry.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        v2rayCore: j['v2rayCore'] as String? ?? 'xray',
        v2rayInboundPort: sanitizePort(j['v2rayInboundPort'], fallback: 10808),
        v2rayEnableMux: j['v2rayEnableMux'] as bool? ?? false,
        v2rayRouteDnsThroughV2Ray: j['v2rayRouteDnsThroughV2Ray'] as bool? ?? true,
        v2rayEnableFragment: j['v2rayEnableFragment'] as bool? ?? true,
        v2rayFragmentPackets: j['v2rayFragmentPackets'] as String? ?? 'tlshello',
        v2rayFragmentLength: j['v2rayFragmentLength'] as String? ?? '100-200',
        v2rayFragmentInterval: j['v2rayFragmentInterval'] as String? ?? '10-20',
        v2rayActiveConfigId: j['v2rayActiveConfigId'] as String? ?? '',
        v2rayConfigs: (j['v2rayConfigs'] as List<dynamic>?)
                ?.map((e) => V2RayConfigEntry.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        shardCustomCfIp: j['shardCustomCfIp'] as String? ?? '',
        shardSmartSplit: j['shardSmartSplit'] as bool? ?? false,
        shardRotateIp: j['shardRotateIp'] as bool? ?? true,
      );

  Map<String, dynamic> toJson() => {
        'theme': theme,
        'language': language,
        'egressRegion': egressRegion,
        'disableTimeouts': disableTimeouts,
        'localSocksProxyPort': localSocksProxyPort,
        'localHttpProxyPort': localHttpProxyPort,
        'useCustomProxyPorts': useCustomProxyPorts,
        'allowLanConnections': allowLanConnections,
        'setSystemProxy': setSystemProxy,
        'autoConnect': autoConnect,
        'startWithWindows': startWithWindows,
        'ipScannerEnabled': ipScannerEnabled,
        'minimizeToTray': minimizeToTray,
        'onCloseAction': onCloseAction,
        'upstreamProxy': upstreamProxy,
        'upstreamProxyEnabled': upstreamProxyEnabled,
        'upstreamProxyScheme': upstreamProxyScheme,
        'upstreamProxyUsername': upstreamProxyUsername,
        'upstreamProxyPassword': upstreamProxyPassword,
        'systemWideTunneling': systemWideTunneling,
        'killSwitchEnabled': killSwitchEnabled,
        'protocolMode': protocolMode,
        'beastMode': beastMode,
        'cdnFrontingCustomIpList': cdnFrontingCustomIpList,
        'cdnFrontingCustomSni': cdnFrontingCustomSni,
        'cdnFrontingSkipCertVerify': cdnFrontingSkipCertVerify,
        'autoFindIpAndSni': autoFindIpAndSni,
        'saveFoundIpsAndSni': saveFoundIpsAndSni,
        'frontedMeekCDNScanBuiltInSets': frontedMeekCDNScanBuiltInSets,
        'establishTunnelTimeoutSeconds': establishTunnelTimeoutSeconds,
        'conduitMode': conduitMode,
        'conduitCompartmentId': conduitCompartmentId,
        'conduitRejectCensoredCountries': conduitRejectCensoredCountries,
        'lanAuthEnabled': lanAuthEnabled,
        'lanProxyUsername': lanProxyUsername,
        'lanProxyPassword': lanProxyPassword,
        'connectionMethod': connectionMethod,
        'aetherProtocol': aetherProtocol,
        'aetherScanMode': aetherScanMode,
        'aetherNoize': aetherNoize,
        'aetherIpVersion': aetherIpVersion,
        'aetherManualPeer': aetherManualPeer,
        'aetherEndpointMasque': aetherEndpointMasque,
        'aetherEndpointWireguard': aetherEndpointWireguard,
        'aetherEndpointWarp': aetherEndpointWarp,
        'aetherEndpointMasqueOnMasque': aetherEndpointMasqueOnMasque,
        'chainedOuterTransport': chainedOuterTransport,
        'chainedPsiphonOuterTransport': chainedPsiphonOuterTransport,
        'chainedTorOuterTransport': chainedTorOuterTransport,
        'chainedSubMode': chainedSubMode,
        'aetherMasqueTransport': aetherMasqueTransport,
        'aetherFragment': aetherFragment,
        'aetherPerfProfile': aetherPerfProfile,
        'aetherMimEnabled': aetherMimEnabled,
        'aetherMimOuter': aetherMimOuter,
        'aetherMimInner': aetherMimInner,
        'aetherMimPeers': aetherMimPeers,
        'aetherQuicV2': aetherQuicV2,
        'aetherMark': aetherMark,
        'torExitCountry': torExitCountry,
        'torBridges': torBridges,
        'splitTunnelEnabled': splitTunnelEnabled,
        'splitTunnelMode': splitTunnelMode,
        'splitTunnelEntries': splitTunnelEntries.map((e) => e.toJson()).toList(),
        'v2rayCore': v2rayCore,
        'v2rayInboundPort': v2rayInboundPort,
        'v2rayEnableMux': v2rayEnableMux,
        'v2rayRouteDnsThroughV2Ray': v2rayRouteDnsThroughV2Ray,
        'v2rayEnableFragment': v2rayEnableFragment,
        'v2rayFragmentPackets': v2rayFragmentPackets,
        'v2rayFragmentLength': v2rayFragmentLength,
        'v2rayFragmentInterval': v2rayFragmentInterval,
        'v2rayActiveConfigId': v2rayActiveConfigId,
        'v2rayConfigs': v2rayConfigs.map((e) => e.toJson()).toList(),
        'shardCustomCfIp': shardCustomCfIp,
        'shardSmartSplit': shardSmartSplit,
        'shardRotateIp': shardRotateIp,
      };

  String encode() => jsonEncode(toJson());

  factory UserSettings.decode(String raw) =>
      UserSettings.fromJson(jsonDecode(raw) as Map<String, dynamic>);

  UserSettings clone() => UserSettings.fromJson(jsonDecode(encode()));
}

class V2RayConfigEntry {
  V2RayConfigEntry({
    required this.id,
    required this.name,
    this.protocol = 'vless',
    required this.address,
    this.port = 443,
    required this.userId,
    this.security = 'reality',
    this.network = 'tcp',
    this.path,
    this.host,
    this.sni,
    this.publicKey,
    this.shortId,
    this.flow,
    this.isActive = false,
    this.latency = -1,
    this.enableFragment,
  });

  String id;
  String name;
  String protocol;
  String address;
  int port;
  String userId;
  String security;
  String network;
  String? path;
  String? host;
  String? sni;
  String? publicKey;
  String? shortId;
  String? flow;
  bool isActive;
  int latency;
  bool? enableFragment;

  factory V2RayConfigEntry.fromJson(Map<String, dynamic> j) => V2RayConfigEntry(
        id: j['id'] as String? ?? '',
        name: j['name'] as String? ?? '',
        protocol: j['protocol'] as String? ?? 'vless',
        address: j['address'] as String? ?? '',
        port: UserSettings.sanitizePort(j['port'], fallback: 443),
        userId: j['userId'] as String? ?? '',
        security: j['security'] as String? ?? 'reality',
        network: j['network'] as String? ?? 'tcp',
        path: j['path'] as String?,
        host: j['host'] as String?,
        sni: j['sni'] as String?,
        publicKey: j['publicKey'] as String?,
        shortId: j['shortId'] as String?,
        flow: j['flow'] as String?,
        isActive: j['isActive'] as bool? ?? false,
        latency: j['latency'] as int? ?? -1,
        enableFragment: j['enableFragment'] as bool?,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'protocol': protocol,
        'address': address,
        'port': port,
        'userId': userId,
        'security': security,
        'network': network,
        'path': path,
        'host': host,
        'sni': sni,
        'publicKey': publicKey,
        'shortId': shortId,
        'flow': flow,
        'isActive': isActive,
        'latency': latency,
        if (enableFragment != null) 'enableFragment': enableFragment,
      };
}
